File: HospitalApp/src/main/java/viewmodel/SceneManager.java
package viewmodel;

import javafx.scene.Scene;
import javafx.stage.Stage;
import viewmodel.panes.*;

/**
 * Handles scene changing logic
 */
public class SceneManager {
    private static final SceneManager ourInstance = new SceneManager();
    private final Scene mainMenuScene;
    private final Scene scanScene;
    private final Scene qrScene;
    private Stage stage;

    private SceneManager() {
        mainMenuScene = new Scene(new MainMenuPane(), Config.WIDTH, Config.HEIGHT);
        scanScene = new Scene(new ScanPane(), Config.WIDTH, Config.HEIGHT);
        qrScene = new Scene(new QRCodePane(), Config.WIDTH, Config.HEIGHT);
        mainMenuScene.getStylesheets().add(Config.CSS_STYLES);
        scanScene.getStylesheets().add(Config.CSS_STYLES);
        qrScene.getStylesheets().add(Config.CSS_STYLES);
    }

    public static SceneManager getInstance() {
        return ourInstance;
    }

    public void setStage(Stage stage) {
        this.stage = stage;
        stage.setOnCloseRequest(e -> ((ScanPane)scanScene.getRoot()).disposeWebCamCamera());
    }

    public QRCodePane getQRCodePane() {
        return (QRCodePane) qrScene.getRoot();
    }

    private void showScene(Scene scene) {
        if (stage == null)
            return;

        stage.setTitle("MediRec");
        stage.hide();
        stage.setScene(scene);
        stage.show();
    }

    public void showMainMenuScene() {
        showScene(mainMenuScene);
    }
    public void showScanScene() {
        showScene(scanScene);
        ((ScanPane)scanScene.getRoot()).startWebCam();
        ((ScanPane)scanScene.getRoot()).setIsSign(false);
    }
    public void showSignScene() {
        showScene(scanScene);
        ((ScanPane)scanScene.getRoot()).startWebCam();
        ((ScanPane)scanScene.getRoot()).setIsSign(true);
    }
    public void showQRScene() { showScene(qrScene); }

}

File: ValidatorApp/src/main/kotlin/viewmodel/SceneManager.kt
package viewmodel

import javafx.scene.Scene
import javafx.stage.Stage
import viewmodel.panes.*

/**
 * Handles scene changing logic
 */
object SceneManager {
    private val logInScene = Scene(LogInPane(), (Config.WIDTH shr 1).toDouble(), (Config.HEIGHT shr 1).toDouble())
    private val mainMenuScene = Scene(MainMenuPane(), (Config.WIDTH).toDouble(), (Config.HEIGHT).toDouble())
    private val scanScene = Scene(ScanPane(), (Config.WIDTH).toDouble(), (Config.HEIGHT).toDouble())
    private val qrScene = Scene(QRCodePane(), (Config.WIDTH).toDouble(), (Config.HEIGHT).toDouble())
    private var stage: Stage? = null

    val qrCodePane: QRCodePane
        get() = qrScene.root as QRCodePane

    init {
        val scenes = arrayOf(logInScene, mainMenuScene, scanScene, qrScene)
        addStylesheets(*scenes)
    }

    private fun addStylesheets(vararg scenes: Scene) {
        for (scene in scenes) {
            scene.stylesheets.add(Config.CSS_STYLES)
        }
    }

    fun setStage(stage: Stage) {
        this.stage = stage
        this.stage!!.title = "MediRec"
        stage.setOnCloseRequest {
            println("Exiting...")
            (scanScene.root as ScanPane).disposeWebCamCamera()
        }
    }

    private fun showScene(scene: Scene) {
        if (stage == null)
            return

        stage!!.hide()
        stage!!.scene = scene
        stage!!.show()
    }

    fun showLogInScene() {
//        showScene(logInScene)
        showMainMenuScene()
    }

    fun showMainMenuScene() {
        showScene(mainMenuScene)
    }

    fun showScanScene(isSign: Boolean) {
        showScene(scanScene)
        (scanScene.root as ScanPane).startWebCam()
        (scanScene.root as ScanPane).setIsSign(isSign)
    }

    fun showQRScene() {
        showScene(qrScene)
    }
}
